src/compat/protocol.rs

```diff
@@ -48,12 +51,40 @@ where
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct CompatProtocolOutbound {
+    pub(crate) protocol_name: &'static [u8],
+    pub(crate) message: CompatMessage,
+}
+
+impl<TSocket> OutboundUpgrade<TSocket> for CompatProtocolOutbound
+where
+    TSocket: AsyncRead + AsyncWrite + Send + Unpin + 'static,
+{
+    type Output = <CompatMessage as OutboundUpgrade<TSocket>>::Output;
+    type Error = <CompatMessage as OutboundUpgrade<TSocket>>::Error;
+    type Future = <CompatMessage as OutboundUpgrade<TSocket>>::Future;
+
+    fn upgrade_outbound(self, socket: TSocket, info: Self::Info) -> Self::Future {
+        CompatMessage::upgrade_outbound(self.message, socket, info)
+    }
+}
+
+impl UpgradeInfo for CompatProtocolOutbound {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(self.protocol_name)
+    }
+}
+
 impl UpgradeInfo for CompatMessage {
     type Info = &'static [u8];
     type InfoIter = iter::Once<Self::Info>;
 
     fn protocol_info(&self) -> Self::InfoIter {
-        iter::once(b"/ipfs/bitswap/1.2.0")
+        unimplemented!()
```

Comment:
This smells bad: why not instead remove this `impl`?

Reply:
It requires more refactoring to remove `UpgradeInfo` impl for `CompatMessage`

Reply:
That refactoring is the required proof that this `unimplemented!()` won’t panic at runtime, though. In other words, I find it reasonable to demand that a production ready library does not contain `unimplemented!()`.

Reply:
That makes sense! However, it's been too long since I made the PR, and I do need some extra time to pick up the context and made the change.